sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,572 @@
+      NULL, // Use HTTP/1.1
+      WINHTTP_NO_REFERER,
+      WINHTTP_DEFAULT_ACCEPT_TYPES, // No media types are accepted by the client
+      WINHTTP_FLAG_SECURE); // Uses secure transaction semantics (SSL/TLS)
+
+  if (!requestHandle)
+  {
+    // Errors include:
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_INVALID_URL
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+    // ERROR_NOT_ENOUGH_MEMORY
+    CleanupHandlesAndThrow(
+        "Error while getting a request handle.", sessionHandle, connectionHandle);
+  }
+
+  std::wstring encodedHeaders;
+  int encodedHeadersLength = 0;
+
+  // TODO: Consider saving this to a field to avoid multiple request processing on retry.
+  auto requestHeaders = request.GetHeaders();
+  std::string requestHeaderString;
+  if (requestHeaders.size() != 0)
+  {
+    // The encodedHeaders will be null-terminated and the length is calculated.
+    encodedHeadersLength = -1;
+    for (auto const& pairs : requestHeaders)
+    {
+      requestHeaderString.append(pairs.first); // string (key)
+      requestHeaderString.append(": ");
+      requestHeaderString.append(pairs.second); // string's value
+      requestHeaderString.append("\r\n");
+    }
+    requestHeaderString.append("\0");
+
+    encodedHeaders = StringToWideString(requestHeaderString);
+  }
+
+  auto streamBody = request.GetBodyStream();
+  int64_t streamLength = streamBody->Length();
+
+  // Send a request.
+  // TODO: For PUT/POST requests, send additional data using WinHttpWriteData.
+  // TODO: Support chunked transfer encoding and missing content-length header.
+  if (!WinHttpSendRequest(
+          requestHandle,
+          requestHeaders.size() == 0 ? WINHTTP_NO_ADDITIONAL_HEADERS : encodedHeaders.c_str(),
+          encodedHeadersLength,
+          WINHTTP_NO_REQUEST_DATA,
+          0,
+          streamLength > 0 ? static_cast<DWORD>(streamLength) : 0,
+          0))
+  {
+    // Errors include:
+    // ERROR_WINHTTP_CANNOT_CONNECT
+    // ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED
+    // ERROR_WINHTTP_CONNECTION_ERROR
+    // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+    // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+    // ERROR_WINHTTP_INTERNAL_ERROR
+    // ERROR_WINHTTP_INVALID_URL
+    // ERROR_WINHTTP_LOGIN_FAILURE
+    // ERROR_WINHTTP_NAME_NOT_RESOLVED
+    // ERROR_WINHTTP_OPERATION_CANCELLED
+    // ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW
+    // ERROR_WINHTTP_SECURE_FAILURE
+    // ERROR_WINHTTP_SHUTDOWN
+    // ERROR_WINHTTP_TIMEOUT
+    // ERROR_WINHTTP_UNRECOGNIZED_SCHEME
+    // ERROR_NOT_ENOUGH_MEMORY
+    // ERROR_INVALID_PARAMETER
+    // ERROR_WINHTTP_RESEND_REQUEST
+    CleanupHandlesAndThrow(
+        "Error while sending a request.", sessionHandle, connectionHandle, requestHandle);
+  }
+
+  // TODO: Allow chunked encoding transfer.
+  if (streamLength > 0 || streamLength == -1)
```

Comment:
OK. Can you give an example of a test case for this?

Reply:
Yes, take a look to this comment: https://github.com/Azure/azure-sdk-for-cpp/issues/249#issuecomment-656997443

Whenever we try to upload more than 256 MB for storage at that REST api version, the server will return an error as soon as it get the first part of the request (Method + headers).

So, you need to check if winHttp client will realize about this and what it will tell you.

One option is that calling `WinHttpSendRequest`  will fail because it would detect that the server return an error response. And if this happen, we should not throw a `TransportException`, we need to parse the Server error and return that as the server response (and do not start the uploading).

Reply:
Based on offline discussion and testing, that issue seems specific to libcurl and doesn't impact winhttp. I will keep an eye out if we see any similar failures that requires adjusting our uploading semantics.